data_steward/retraction/retract_deactivated_pids.py

```diff
@@ -0,0 +1,441 @@
+FROM `{deactivated_pids_project}.{deactivated_pids_dataset}.{deactivated_pids_table}`
+WHERE person_id = {pid} ) ELSE CASE WHEN {end_date_column} IS NULL THEN {start_date_column} >= (
+SELECT deactivated_date
+FROM `{deactivated_pids_project}.{deactivated_pids_dataset}.{deactivated_pids_table}`
+WHERE person_id = {pid}) END END))
+"""
+# Queries to truncate existing tables to remove deactivated EHR PIDS, two different queries for
+# tables with standard entry dates vs. tables with start and end dates
+CLEAN_QUERY_DATE = """
+SELECT *
+FROM `{project}.{dataset}.{table}`
+WHERE person_id != {pid}
+AND {date_column} < (SELECT deactivated_date
+FROM `{deactivated_pids_project}.{deactivated_pids_dataset}.{deactivated_pids_table}`
+WHERE person_id = {pid})
+"""
+CLEAN_QUERY_END_DATE = """
+SELECT *
+FROM `{project}.{dataset}.{table}`
+WHERE person_id != {pid}
+AND (CASE WHEN {end_date_column} IS NOT NULL THEN {end_date_column} < (SELECT deactivated_date
+FROM `{deactivated_pids_project}.{deactivated_pids_dataset}.{deactivated_pids_table}`
+WHERE person_id = {pid} ) ELSE CASE WHEN {end_date_column} IS NULL THEN {start_date_column} < (
+SELECT deactivated_date
+FROM `{deactivated_pids_project}.{deactivated_pids_dataset}.{deactivated_pids_table}`
+WHERE person_id = {pid}) END END)
+"""
+# Deactivated participant table fields to query off of
+PID_TABLE_FIELDS = [{
+    "type": "integer",
+    "name": "person_id",
+    "mode": "required",
+    "description": "The person_id to retract data for"
+}, {
+    "type": "date",
+    "name": "deactivated_date",
+    "mode": "required",
+    "description": "The deactivation date to base retractions on"
+}]
+
+
+def get_pids_datasets_and_tables(project_id):
+    """
+    Get list of tables that have a person_id column; along with a separate
+    list of the datasets containing pids
+    """
+    dataset_obj = bq.list_datasets(project_id)
+    datasets = [d.dataset_id for d in dataset_obj]
+
+    pids_table_list = []
+    pids_dataset_list = []
+    for dataset in datasets:
+        pid_tables = sandbox_and_remove_pids.get_tables_with_person_id(
+            project_id, dataset)
+        if pid_tables:
+            pids_table_list.extend(pid_tables)
+            pids_dataset_list.append(dataset)
+
+    return pids_table_list, pids_dataset_list
+
+
+def get_date_info_for_pids_tables(project_id):
+    """
+    Loop through tables within all datasets and determine if the table has an end_date date or a date field. Filtering
+    out person and death table and keeping only tables with PID and an upload or start/end date associated.
+
+    :param project_id: bq name of project_id
+    :return: filtered dataframe which includes the following columns for each table in each dataset with a person_id
+    'project_id', 'dataset_id', 'table', 'date_column', 'start_date_column', 'end_date_column'
+    """
+    logging.info('getting all tables with person id in %s' % project_id)
```

Comment:
I think the consensus from the last engineer sync meeting was to use f-strings as part of logging.  Please make this change.